### PR TITLE
Add platform message queue

### PR DESCRIPTION
**Pull request by mravn-google**
Introduces a non-null default value of the `Window.onPlatformMessage` callback that queues incoming messages. This replaces null as default value which caused early messages to be lost. The queue is flushed into a custom message handler when set.

In a Flutter app, the custom message handler is an implementation defined in `BinaryMessages`. That handler will  asynchronously forward messages to registered message handlers according to channel name.

The asynchronous forwarding inside `BinaryMessages` implies that handlers registered with `BinaryMessages` during the first widget tree build will receive any queued messages, even though that build happens *after* `Window.onPlatformMessage` is set and the queue is flushed.

The correct delivery of early messages thus depends critically on current async processing. A `flutter/flutter` test will be added on engine roll to verify this behavior.

Partial fix of https://github.com/flutter/flutter/issues/8711

### CODE REVIEW

File: lib/ui/window.dart [hunk 1]
```diff
@@ -283,7 +283,24 @@ class Window {
...
+  /// The default handler queues incoming messages and relays them to
+  /// the first custom handler set, flushing the queue. The default handler is
+  /// restored, if [onPlatformMessage] is set to null.
+  PlatformMessageCallback get onPlatformMessage => _onPlatformMessage;
+  void set onPlatformMessage(PlatformMessageCallback callback) {
+    if (callback == null) {
+      _onPlatformMessage = _queueingPlatformMessageCallback;
+    } else {
+      if (_onPlatformMessage == _queueingPlatformMessageCallback) {
+        for (_PlatformMessage message in _messageQueue)
```
Comment (abarth) [inline lib/ui/window.dart]: Consider what happens if the `onPlatformMessage` setter is called during this iteration.  You probably want to make a local copy of the queue and clear it out before sending callbacks.

Also, calling the callback re-entrantly here makes a difficult-to-use API.  This setter is likely to be called some time during startup before all the receivers are fully registered.  Consider delaying deliver until a microtask.  (If you do that, you probably want to freeze the set of messages to be delivered when you schedule the microtask and consider the case where the setter is returned to null between the time the task is scheduled and when it executes.)

File: lib/ui/window.dart [hunk 2]
```diff
@@ -294,3 +311,21 @@ class Window {
...
+  _PlatformMessage(this.name, this.data, this.responseCallback);
+}
+
+/// Queue of unhandled [_PlatformMessage]s.
+final List<_PlatformMessage> _messageQueue = <_PlatformMessage>[];
+
+/// Queueing platform message handler.
+final PlatformMessageCallback _queueingPlatformMessageCallback =
+  (String name, ByteData data, PlatformMessageResponseCallback responseCallback) {
+    _messageQueue.add(new _PlatformMessage(name, data, responseCallback));
```
Comment (abarth) [inline lib/ui/window.dart]: It looks like this buffer will grow without bound if no one ever registers an onPlatformMessage callback.

**Comment by abarth** [comment]
> The asynchronous forwarding inside ...

The `dart:ui` library should not depend on details like this about `package:flutter`.  They are independent layers of the system.  The `dart:ui` library should have a sensible interface on its own.

File: lib/ui/window.dart [hunk 3]
```diff
@@ -283,7 +283,24 @@ class Window {
   /// Message handlers must call the function given in the `callback` parameter.
   /// If the handler does not need to respond, the handler should pass `null` to
   /// the callback.
-  PlatformMessageCallback onPlatformMessage;
+  ///
+  /// The default handler queues incoming messages and relays them to
+  /// the first custom handler set, flushing the queue. The default handler is
+  /// restored, if [onPlatformMessage] is set to null.
```
Comment (Hixie) [inline lib/ui/window.dart]: nit: remove the comma

File: lib/ui/window.dart [hunk 4]
```diff
@@ -294,3 +311,21 @@ class Window {
...
+/// Platform message for queueing.
+class _PlatformMessage {
+  final String name;
+  final ByteData data;
+  final PlatformMessageResponseCallback responseCallback;
+
+  _PlatformMessage(this.name, this.data, this.responseCallback);
+}
+
+/// Queue of unhandled [_PlatformMessage]s.
```
Comment (Hixie) [inline lib/ui/window.dart]: nit: this is a List, not a Queue

**Comment by Hixie** [comment]
An alternative approach would be to require all plugins to react to a "report your current state" signal. Then, when you register the handler, you can send that signal to all the plugins. That would avoid queuing any messages, so the queue wouldn't grow without bounds. It would handle the case of handing over the entire system from one framework to another, since the second framework could repeat that signal to get up to speed (right now to do that you'd have to somehow hand over all the state from one framework to the other, which is a lot of fiddly work). It would also remove any concern over changing the handler midway through the flushing of the queue.

**Comment by mravn-google** [comment]
Thanks for looking into this.

**Old plan**: add a simple fix to avoid dropping platform messages sent on app launch (e.g. for https://github.com/goposse/fluro routing on Android `onResume` from background state, cf. https://github.com/flutter/flutter/issues/8711#issuecomment-304681212).
**New plan**: leave initial routing to https://github.com/flutter/flutter/issues/10399 and instead follow Ian's lead on a protocol to allow the framework to signal plugins that they should report their current state to the Dart side. This mechanism seems capable of also addressing https://github.com/flutter/flutter/issues/10437.

Closing this PR.

**Comment by mravn-google** [comment]
Update: https://github.com/flutter/flutter/issues/8711#issuecomment-305744394

**Pull request closed without merging**